libvast/src/directory.cpp

```diff
@@ -128,32 +128,32 @@ caf::expected<size_t> recursive_size(const std::filesystem::path& root_dir) {
   return total_size;
 }
 
-std::vector<path>
-filter_dir(const path& dir, std::function<bool(const path&)> filter,
+caf::expected<std::vector<std::filesystem::path>>
+filter_dir(const std::filesystem::path& root_dir,
+           std::function<bool(const std::filesystem::path&)> filter,
            size_t max_recursion) {
-  std::vector<path> result;
-  if (max_recursion == 0)
-    return result;
-  for (auto& f : directory(dir)) {
-    switch (f.kind()) {
-      default: {
-        if (!filter || filter(f))
-          result.push_back(f);
-        break;
-      }
-      case path::directory: {
-        // Recurse directories depth-first.
-        auto paths = filter_dir(f, filter, --max_recursion);
-        if (!paths.empty()) {
-          auto begin = std::make_move_iterator(paths.begin());
-          auto end = std::make_move_iterator(paths.end());
-          result.insert(result.end(), begin, end);
-          std::sort(result.begin(), result.end());
-        }
-        break;
-      }
-    }
+  std::vector<std::filesystem::path> result;
+  std::error_code ec{};
```

Comment:
```suggestion
  std::error_code ec;
```

Herb Sutter advocates to use `auto ec = std::error_code{}`. I don't think we have established a convention in our code base. @tenzir/backend any thoughts on what to go with? I am only asking because we usually don't write braces in a statement like `T x{}`.

Reply:
We should always default-initialize these to avoid bugs. So one of the following two, always:

```
std::error_code ec{};
auto ec = std::error_code{};
```

Which of these you use comes down to personal preference.

Reply:
Right, I always recommend default initializing `std::error_code` when using them in standard library API calls as @dominiklohmann pointed out. More generally, it's just a habit of my coding style in what I use day-to-day at work.  We prefer almost always auto like you said @mavam, but we often brace initialize everything at declaration time to avoid uninitialized variables and having to think about it.